cis194/week2/zehua/src/Exercises.hs

```diff
@@ -31,3 +32,50 @@ parseMessage s =
 
 parse :: String -> [LogMessage]
 parse = fmap parseMessage . lines
+
+-- ex2
+insert :: LogMessage -> MessageTree -> MessageTree
+insert (Unknown _) t = t
+insert lm Leaf = Node Leaf lm Leaf
+insert lm@(LogMessage _ ts _ ) (Node l lm1@(LogMessage _ ts1 _) r)
+  | ts < ts1 = Node (insert lm l) lm1 r
+  | otherwise = Node l lm1 (insert lm r)
```

Comment:
compiler complains about `Unknown` being a possible value in the node, which is not really possible if the `MessageTree` is an output from `insert`. It would be a runtime error if someone passes in such a `MessageTree` as the initial value.
```
/Users/zehualiu/src/mine/2019-winter-Haskell-school/cis194/week2/zehua/src/Exercises.hs:38:1: warning: [-Wincomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘insert’:
        Patterns not matched: (LogMessage _ _ _) (Node _ (Unknown _) _)
   |
38 | insert (Unknown _) t = t
   | ^^^^^^^^^^^^^^^^^^^^^^^^...
```

Reply:
If we remove the pattern `insert (Unknown _) t = t` and add below as the last pattern:
```Haskell
insert _ _ -> t
```
Then we will not have this `non-exhaustive` warning. Also, I think this is a sensible logic.

A note on `non-exhaustive pattern match` warning: They should not be ignored, because it essentially says out function is partial.

Reply:
added `insert _ t = t`. still need `insert (Unknown _) t = t` at the beginning so that the `Leaf` case will not kick in first.